File: 0042-trapping-rain-water/0042-trapping-rain-water.java
class Solution {
    public int trap(int[] height) {
        int maxl = 0;
        int maxr = 0;
        int ans = 0;
        int l[] = new int[height.length];
        int r[] = new int[height.length];
        for (int i = 0, j = height.length - 1; i < height.length; i++, j--) {
            maxl = Math.max(maxl, height[i]);
            l[i] = maxl;
            maxr = Math.max(maxr, height[j]);
            r[j] = maxr;
        }
        for (int i = 0; i < height.length; i++) {
            ans += Math.min(l[i], r[i]) > height[i] ? Math.min(l[i], r[i]) - height[i] : 0;
        }
        return ans;
    }
}

File: 0042-trapping-rain-water/0042-trapping-rain-water.py
class Solution:
    def trap(self, height: List[int]) -> int:
        l, r = 0, len(height) - 1
        maxL, maxR = 0, 0
        ans = 0
        while l <= r:
            if maxL <= maxR:
                if maxL - height[l] > 0:
                    ans += maxL - height[l]
                maxL = max(maxL, height[l])
                l += 1
            else:
                if maxR - height[r] > 0:
                    ans += maxR - height[r]
                maxR = max(maxR, height[r])
                r -= 1
        return ans